tests/images/kafka.rs

```diff
@@ -0,0 +1,147 @@
+use futures::{
+    future::{self, join_all},
+    StreamExt, TryStreamExt,
+};
+use rdkafka::{
+    config::ClientConfig,
+    consumer::{stream_consumer::StreamConsumer, Consumer},
+    producer::{FutureProducer, FutureRecord},
+    Message,
+};
+use std::{collections::HashMap, time::Duration};
+use uuid::Uuid;
+
+use testcontainers::{
+    clients,
+    core::{RunArgs, WaitFor},
+    images::{kafka, zookeeper},
+    Image,
+};
+
+const ZOOKEEPER_PORT: u16 = 2181;
+
+/// Populates `topic` once a Kafka broker is up and running.
+fn create_topic(docker: &clients::Cli, network: &Uuid, topic: &str) {
+    let topic_creation_command = format!("cub kafka-ready -b broker:{} 1 20 && \
+kafka-topics --create --if-not-exists --bootstrap-server broker:{} --partitions 1 --replication-factor 1 --topic {}",
+	kafka::DEFAULT_BROKER_LISTENER_PORT, kafka::DEFAULT_BROKER_LISTENER_PORT, topic);
+
+    // Once `topic` has been created, it is ok to destroy the container at the end of the function
+    // because it is meant to be a one-shot.
+    docker.run_with_args(
+        kafka::Kafka::default()
+            .with_wait_for(WaitFor::message_on_stdout("Created topic"))
+            .with_args(kafka::KafkaArgs(vec![
+                "/bin/bash".to_owned(),
+                "-c".to_owned(),
+                topic_creation_command,
+            ])),
+        RunArgs::default()
+            .with_name("kafka-topic-creator")
+            .with_network(network),
+    );
+}
+
+#[tokio::test]
+async fn test_produce_and_consume_messages() {
+    // The crate does not seem to support Docker Compose at this time.
+    // We run ZooKeeper and Kafka containers separately, with the ZooKeeper
+    // container running first because the Kafka container depends on it.
+
+    let docker = clients::Cli::default();
+    let network = Uuid::new_v4();
+
+    let _zk_node = docker.run_with_args(
+        zookeeper::Zookeeper::default(),
+        RunArgs::default()
+            .with_name("zookeeper")
+            .with_network(network),
+    );
+
+    let mut env_vars = HashMap::new();
+    env_vars.insert(
+        "KAFKA_ZOOKEEPER_CONNECT".to_owned(),
+        format!("zookeeper:{}", ZOOKEEPER_PORT),
+    );
+    let kafka_broaker_node = docker.run_with_args(
+        kafka::Kafka::default()
+            .with_env_vars(env_vars)
+            .with_wait_for(WaitFor::message_on_stdout("started")),
+        RunArgs::default()
+            .with_name("broker")
+            .with_network(network)
+            .with_mapped_port((
+                kafka::DEFAULT_LOCALHOST_LISTENER_PORT,
+                kafka::DEFAULT_LOCALHOST_LISTENER_PORT,
+            )),
```

Comment:
So this is something I could not hide from the user completely after my best effort. When the user runs a Kafka container it will look like the following:
```
let docker = clients::Cli::default();
let kafka_node = docker.run_with_args(
    kafka::Kafka::default(),
    RunArgs::default().with_mapped_port((9093, 9093)),
);
```
The reason for the port mapping is that before the container runs, one of the required environment variables `KAFKA_ADVERTISED_LISTENERS` needs to specify the exposed port for a Kafka port but this exposed port is only known after the container has run. To work around this problem, we need to specify the port mapping so that the exposed port is known before the container runs.

I have noticed that other `testcontainers` client libraries seem to provide a way to obtain the exposed port for a given port (an [example](https://github.com/testcontainers/testcontainers-java/blob/master/modules/kafka/src/main/java/org/testcontainers/containers/KafkaContainer.java#L56) in Java, [another](https://github.com/testcontainers/testcontainers-python/blob/bc969c53680606d7fff6a2e8350f5a6f234fcd38/testcontainers/kafka.py#L35) in Python)? If we can do the same in Rust we should be able to remove the port mapping completely.

Reply:
I assume `KAFKA_ADVERTISED_LISTENERS` is only used internally within the container right?

The port within the container can always be 9093. This env variable should not matter for how docker maps the port onto the host system.

Reply:
Please see https://github.com/testcontainers/testcontainers-rs/pull/282#discussion_r612841110